feat(qrcode): support click & iconSize configuration

thinkasany opened the pull request:
<!--
First of all, thank you for your contribution! 😄
For requesting to pull a new feature or bugfix, please send it from a feature/bugfix branch based on the `master` branch.
Before submitting your pull request, please make sure the checklist below is confirmed.
Your pull requests will be merged after one of the collaborators approve.
Thank you!
-->

[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
close #49214
<!--
1. Put the related issue or discussion links here.
2. close #xxxx or fix #xxxx for instance.
-->

### 💡 Background and solution
![image](https://github.com/ant-design/ant-design/assets/117748716/a9e82839-2804-46e6-bfb2-c361425de4ae)<!--
1. Describe the problem and the scenario.
2. GIF or snapshot should be provided if includes UI/interactive modification.
3. How to fix the problem, and list the final API implementation and usage sample if that is a new feature.
-->

### 📝 Changelog

<!--
Describe changes from the user side, and list all potential break changes or other risks.
--->

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |         feat(qrcode): support click & iconSize configuration  |
| 🇨🇳 Chinese |    feat(qrcode): 支持点击事件和iconSize宽高的配置       |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

thinkasany committed:
feat(qrcode): support click & iconSize configuration

li-jia-nan commented on components/qr-code/__tests__/index.test.tsx:
```diff
@@ -46,6 +46,13 @@ describe('QRCode test', () => {
+  it('support click', () => {
+    const handleClick = jest.fn();
+    const { container } = render(<QRCode value="test" onClick={handleClick} />);
+    fireEvent.click(container?.querySelector<HTMLButtonElement>('.ant-qrcode')!);
```
```suggestion
    fireEvent.click(container?.querySelector<HTMLDivElement>('.ant-qrcode')!);
```

li-jia-nan commented on components/qr-code/index.en-US.md:
```diff
@@ -41,7 +41,7 @@ Common props ref：[Common props](/docs/react/common-props)
 | icon | include image url (only image link are supported) | string | - |
 | size | QRCode size | number | 160 |
-| iconSize | include image size | number | 32 |
+| iconSize | include image size | number \| { width: number; height: number } | 32 | 5.18.0 |
```
```suggestion
| iconSize | include image size | number \| { width?: number; height?: number } | 32 | 5.19.0 |
```

li-jia-nan commented on components/qr-code/index.en-US.md:
```diff
@@ -41,7 +41,7 @@ Common props ref：[Common props](/docs/react/common-props)
 | icon | include image url (only image link are supported) | string | - |
 | size | QRCode size | number | 160 |
-| iconSize | include image size | number | 32 |
+| iconSize | include image size | number \| { width: number; height: number } | 40 | 5.18.0 |
```
```suggestion
| iconSize | include image size | number \| { width?: number; height?: number } | 40 | 5.19.0 |
```

li-jia-nan commented on components/qr-code/index.tsx:
```diff
@@ -41,8 +42,8 @@ const QRCode: React.FC<QRCodeProps> = (props) => {
     y: undefined,
-    height: iconSize,
-    width: iconSize,
+    height: typeof iconSize === 'number' ? iconSize : iconSize?.height,
```
这个逻辑可以抽成一个方法 mergedSize

thinkasany replied:
感觉好像这边场景还挺简单的，不需要抽出去，undefined / null  也能正常显示

![image](https://github.com/ant-design/ant-design/assets/117748716/dee9d9f3-fc37-409d-b46b-b12a757afeb6)

li-jia-nan commented on components/qr-code/interface.ts:
```diff
@@ -29,9 +29,10 @@ export interface QRCodeProps extends QRProps {
   prefixCls?: string;
   icon?: string;
-  iconSize?: number;
+  iconSize?: number | { width: number; height: number };
```
```suggestion
  iconSize?: number | { width?: number; height?: number };
```

thinkasany replied:
qrcode 这边的ImageSettings的width height 不支持可选，会出现一些错误，要不然还是严格一点不支持可选？

![image](https://github.com/ant-design/ant-design/assets/117748716/78c5b700-9f1e-4593-b792-e5314aebda15)

![image](https://github.com/ant-design/ant-design/assets/117748716/eb37b473-bc78-4ae5-8ccc-0338c413e59f)

li-jia-nan replied:
把上面的默认值40放到下面：

```ts
const imageSettings: QRProps['imageSettings'] = {
  src: icon,
  x: undefined,
  y: undefined,
  height: typeof iconSize === 'number' ? iconSize : iconSize?.width ?? 40,
  width: typeof iconSize === 'number' ? iconSize : iconSize?.height ?? 40,
  excavate: true,
};
```

li-jia-nan commented on components/qr-code/interface.ts:
```diff
@@ -29,9 +29,10 @@ export interface QRCodeProps extends QRProps {
   errorLevel?: 'L' | 'M' | 'Q' | 'H';
   status?: 'active' | 'expired' | 'loading' | 'scanned';
   onRefresh?: () => void;
+  onClick?: () => void;
```
```suggestion
  onClick?: React.MouseEventHandler<HTMLDivElement>;
```

li-jia-nan commented on components/qr-code/index.tsx:
```diff
@@ -86,7 +87,7 @@ const QRCode: React.FC<QRCodeProps> = (props) => {
 
   return wrapCSSVar(
-    <div className={mergedCls} style={mergedStyle}>
+    <div className={mergedCls} style={mergedStyle} onClick={onClick}>
```
这个 onClick 不确定挂在外层，还是挂在里面的 `Canvas` or `Svg` 元素，感觉需要讨论一下

afc163 replied:
放外面吧，直接把 rest 都传进来，这样也能解决 Tooltip 等无法生效的问题。

li-jia-nan replied:
有道理

li-jia-nan replied:
那 interface 里面就不用加 onClick 了，直接把 React.HTMLAttributes<HTMLDivElement> 合并进来吧

thinkasany committed:
fix

fix

use rest

union  React.HTMLAttributes<HTMLElement

fix ImageSettings interface

type fix

add default 40

clear unused

li-jia-nan commented on components/qr-code/index.tsx:
```diff
@@ -13,7 +13,7 @@ import { useToken } from '../theme/internal';
 import useStyle from './style/index';
 
-const QRCode: React.FC<QRCodeProps> = (props) => {
+const QRCode: React.FC<QRCodeProps & React.HTMLAttributes<HTMLElement>> = (props) => {
```
这里不对，不应该改这个，应该直接改 QRCodeProps：

```ts
export interface QRCodeProps extends QRProps, React.HTMLAttributes<HTMLDivElement> {

}
```

li-jia-nan commented on components/qr-code/index.tsx:
```diff
@@ -86,7 +87,7 @@ const QRCode: React.FC<QRCodeProps> = (props) => {
 
   return wrapCSSVar(
-    <div className={mergedCls} style={mergedStyle}>
+    <div className={mergedCls} style={mergedStyle} {...rest}>
```
rest 顺序不对，应该放在前面，否则会覆盖 className 和 style

li-jia-nan commented on components/qr-code/index.tsx:
```diff
@@ -41,8 +42,8 @@ const QRCode: React.FC<QRCodeProps> = (props) => {
     y: undefined,
-    height: iconSize,
-    width: iconSize,
+    height: typeof iconSize === 'number' ? iconSize : iconSize?.height ?? 40,
```
上面解构的默认值 40 可以去掉了

thinkasany replied:
done. thx~

wanpan11 commented:
更新下分支看看现在应该不会有 diff 才对 😂

thinkasany commented:
> 你这个对比的快照有点老 更新下分支看看现在应该不会有 diff 才对 😂

rebase 过 upstream/feature了，如果落后分支了这边可以看到这个，但是目前没有这个。不知道是不是feature还没同步了master的代码还是什么其他的原因。 
![image](https://github.com/ant-design/ant-design/assets/117748716/81780aaf-c539-48f5-9d3b-fe2872310bd0)

thinkasany committed:
empty commit

afc163 committed:
Merge branch 'feature' into feat-qrcode
